pkg/controller/deployment/sync.go

```diff
@@ -43,7 +44,70 @@ func (dc *DeploymentController) syncStatusOnly(deployment *extensions.Deployment
 	}
 
 	allRSs := append(oldRSs, newRS)
-	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
+	newStatus, err := dc.calculateStatus(allRSs, newRS, deployment)
+	if err != nil {
+		return err
+	}
+
+	if reflect.DeepEqual(deployment.Status, newStatus) {
+		return nil
+	}
+
+	newDeployment := deployment
+	newDeployment.Status = newStatus
+	_, err = dc.client.Extensions().Deployments(newDeployment.Namespace).UpdateStatus(newDeployment)
+	return err
+}
+
+// hasTimedOut let us know if a deployment did not manage to finish before the specified
+// timeout (progressDeadlineSeconds). The controller should not reconcile deployments that
+// timed out. In a future PR we will add automatic rollback, then the controller will need
+// to maintain the last replica set that was successfully deployed instead of the newest.
+func (dc *DeploymentController) hasTimedOut(deployment *extensions.Deployment) (bool, error) {
+	if deployment.Spec.ProgressDeadlineSeconds == nil || deployment.Spec.RollbackTo != nil {
+		return false, nil
+	}
+
+	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(deployment, false)
+	if err != nil {
+		return false, err
+	}
+	allRSs := oldRSs
+
+	// There is a template change so we don't need to check for any progress right now.
+	if newRS == nil {
+		return false, nil
+	}
+	allRSs = append(allRSs, newRS)
+
+	switch {
+	case deploymentutil.IsReplicaSetComplete(newRS):
+		// If the latest replica set has been tagged as complete then there is no need
+		// to check for the deployment progress anymore.
+		return false, nil
+
+	case deploymentutil.IsReplicaSetFailed(newRS):
+		// If the latest replica set has been tagged as failed then we already know about
+		// the lack of progress of this deployment.
+		return true, dc.syncDeploymentStatus(allRSs, newRS, deployment)
```

Comment:
Yes, infant mortality detection is a different issue: https://github.com/kubernetes/kubernetes/issues/18568